### PR TITLE
Fix `getDocPage()` exposing parser-owned structures by reference

### PR DESCRIPTION
**Pull request by dahlia**
## Summary

`getDocPage()` reuses parser-owned `usage` terms and `getDocFragments()` message/entry objects instead of cloning them. This means mutating the returned `DocPage` or the `defaultUsageLine` argument passed to a command's `usageLine` callback can corrupt the parser definition and change subsequent help/documentation output.

The root cause is in `buildDocPage()` inside *packages/core/src/parser.ts*, which has three reference-leak vectors:

1. `const usage = [...normalizeUsage(parser.usage)]` only shallow-copies the top-level array; the individual `UsageTerm` objects inside remain parser-owned references.
2. `DocEntry` objects from `parser.getDocFragments()` are pushed directly into section arrays, and `brief`, `description`, `footer` messages are passed through to the returned `DocPage` unchanged.
3. `usage.slice()` passed as `defaultUsageLine` to `usageLine` callbacks creates a new array, but the elements are still references, so callback-side mutation flows back into parser-owned structures.

For example, this sequence corrupts the parser:

```typescript
const page1 = getDocPageSync(parser);
(page1.usage[0] as any).name = "MUTATED";
(page1.brief[0] as any).text = "MUTATED";

const page2 = getDocPageSync(parser);
// page2 now contains the mutated values
```

## Changes

This PR adds a set of deep-clone utilities and applies them in `buildDocPage()` to isolate the returned `DocPage` from parser-owned state.

New public APIs:

- `cloneUsageTerm()` and `cloneUsage()` in `@optique/core/usage`: recursively deep-clone `UsageTerm` trees. For `command` terms, a function-valued `usageLine` is preserved by reference (functions are stateless callbacks), while an array-valued `usageLine` is deep-cloned.
- `cloneMessageTerm()` and `cloneMessage()` in `@optique/core/message`: deep-clone `MessageTerm` arrays without `structuredClone`, which throws `DataCloneError` on Node.js when a `url()` term embeds a `URL` object. The `url` variant is cloned by constructing a new `URL` from `href`.
- `cloneDocEntry()` in `@optique/core/doc`: composes `cloneUsageTerm` for the entry term and `cloneMessage` for `description`, `default`, and `choices`.

In `buildDocPage()`, seven sites are updated:

- Usage array construction uses `cloneUsage(normalizeUsage(parser.usage))` instead of a shallow spread.
- All three entry-push sites in the section-building loop apply `cloneDocEntry()`.
- Both `defaultUsageLine` slices passed to `usageLine` callbacks are wrapped in `cloneUsage()`.
- `brief`, `description`, and `footer` in the return object use `cloneMessage()`.

## Test plan

- [ ] Unit tests for `cloneUsageTerm` / `cloneUsage` covering all 9 `UsageTerm` variants, including function-valued vs array-valued `usageLine` on command terms
- [ ] Unit tests for `cloneMessageTerm` / `cloneMessage` covering `url`, `optionNames`, `values`, and primitive-only variants
- [ ] Unit tests for `cloneDocEntry` with all fields present, optional fields absent, and a URL-containing description
- [ ] Regression tests in *parser.test.ts* verifying that mutating a returned `DocPage`'s usage, brief, and entry descriptions does not affect subsequent `getDocPageSync()` calls, and that `usageLine` callback mutations do not corrupt `parser.usage`
- [ ] Full test suite passes across Deno, Node.js, and Bun (`mise test`)

Closes https://github.com/dahlia/optique/issues/500

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -748,6 +748,21 @@ To be released.
...
+
+ -  Added deep-clone utilities for parser structures:
+
+     -  `cloneUsageTerm()` and `cloneUsage()` in `@optique/core/usage`
+     -  `cloneDocEntry()` in `@optique/core/doc`
+     -  `cloneMessageTerm()` and `cloneMessage()` in `@optique/core/message`
+
+    These are used internally by `getDocPage()` to isolate returned pages
+    from parser-owned state, and are also available as public APIs.
+    [[#500], [#697]]
```
[the remark this answers is not in the document]
Comment (dahlia) [reply CHANGES.md:771]: Fixed. Added user-facing context to the changelog entry while keeping the API enumeration, since both are useful. See 181c71e.

File: packages/core/src/doc.ts [hunk 2]
```diff
@@ -121,6 +123,30 @@ export interface DocFragments {
...
 }
 
+/**
+ * Creates a deep clone of a {@link DocEntry}.  The `term` is cloned via
+ * {@link cloneUsageTerm}, and `description`, `default`, and `choices`
+ * messages are cloned via `structuredClone`.
+ *
+ * @param entry The documentation entry to clone.
+ * @returns A structurally equal but referentially distinct copy.
+ * @since 1.0.0
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/doc.ts:133]: Fixed. Updated the JSDoc to reference `cloneMessage()` instead of `structuredClone`. See 849b1e3.

File: packages/core/src/message.ts [hunk 3]
```diff
@@ -160,6 +160,41 @@ export type MessageTerm =
...
+
+/**
+ * Creates a deep clone of a {@link Message} array and all of its terms.
+ *
+ * @param msg The message to clone.
+ * @returns A structurally equal but referentially distinct copy.
+ * @since 1.0.0
+ */
+export function cloneMessage(msg: Message): MessageTerm[] {
+  return msg.map(cloneMessageTerm);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/message.ts:195]: Fixed. `cloneMessage()` now returns `Message` (the readonly alias) instead of `MessageTerm[]`. See a319508.

File: packages/core/src/parser.test.ts [hunk 4]
```diff
@@ -2656,3 +2658,103 @@ describe("getDocPage: section merging (issue #138)", () => {
...
+      page2.sections.length > 0 &&
+      page2.sections[0].entries.length > 0 &&
+      page2.sections[0].entries[0].description != null
+    ) {
+      assert.notEqual(
+        (page2.sections[0].entries[0].description[0] as { text?: string })
+          .text,
+        "MUTATED",
+      );
+    }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/parser.test.ts]: Fixed. Replaced conditional guards with explicit `assert.ok()` preconditions so the test fails fast if the page structure changes unexpectedly. See 74adeff.

File: packages/core/src/parser.ts [hunk 5]
```diff
@@ -1043,11 +1054,13 @@ function buildDocPage(
...
     }
-    const defaultUsageLine = usage.slice(usageIndex + 1);
+    const defaultUsageLine = cloneUsage(usage.slice(usageIndex + 1));
     const customUsageLine = typeof term.usageLine === "function"
       ? term.usageLine(defaultUsageLine)
       : term.usageLine;
-    const normalizedCustomUsageLine = normalizeUsage(customUsageLine);
+    const normalizedCustomUsageLine = cloneUsage(
+      normalizeUsage(customUsageLine),
+    );
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/parser.ts:1063]: I'll leave this as-is. The two call sites have slightly different contexts (one uses `usageIndex + 1`, the other uses `1`) and the logic is only three lines each. Extracting a helper would add indirection without meaningful deduplication.

File: packages/core/src/doc.test.ts [hunk 6]
```diff
@@ -2614,3 +2616,66 @@ describe("branch coverage: doc.ts edge cases", () => {
...
+  });
+
+  it("should not leak mutations back to the original", () => {
+    const entry: DocEntry = {
+      term: { type: "argument", metavar: "FILE" },
+      description: [{ type: "text", text: "original" }],
+    };
+    const cloned = cloneDocEntry(entry);
+    (cloned.description as unknown as Record<string, unknown>[])[0].text =
+      "mutated";
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/doc.test.ts]: Fixed. Replaced the `as unknown as` double cast with `// @ts-expect-error` to match the project's test convention. See 63d3f65.

File: packages/core/src/parser.test.ts [hunk 7]
```diff
@@ -2656,3 +2658,91 @@ describe("getDocPage: section merging (issue #138)", () => {
...
+    const page1 = getDocPageSync(parser as Parser<"sync", unknown, unknown>);
+    assert.ok(page1);
+
+    // Mutate the returned page's usage
+    if (page1.usage != null && page1.usage.length > 1) {
+      (page1.usage[1] as Record<string, unknown>).type = "MUTATED";
+    }
+
+    // The shared array should not be corrupted
+    assert.equal(sharedUsage[0].type, "ellipsis");
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/parser.test.ts:2747]: Fixed. Added explicit precondition asserts (`page1.usage[1].type === "ellipsis"`) before the mutation so the test fails fast if the usageLine override is not applied. See 63d3f65.

File: packages/core/src/parser.ts [hunk 8]
```diff
@@ -1080,19 +1093,22 @@ function buildDocPage(
...
   }
   return {
     usage,
     sections,
-    ...(brief != null && { brief }),
-    ...(description != null && { description }),
-    ...(footer != null && { footer }),
+    ...(brief != null && { brief: cloneMessage(brief) }),
+    ...(description != null && { description: cloneMessage(description) }),
+    ...(footer != null && { footer: cloneMessage(footer) }),
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/parser.ts:1111]: This is a false positive. `examples`, `author`, and `bugs` are fields on `DocPage` but not on `DocFragments`. `buildDocPage()` constructs its return value from `parser.getDocFragments()`, which returns `DocFragments` — those fields are never present in the source data, so there is nothing to propagate or clone.
